---
title: "circRNAs"
author: ""
date: "14/12/2020"
output: 
  html_document: 
      toc: TRUE
      toc_float: TRUE
      code_folding: hide
  pdf_document:
      keep_tex: TRUE
---


**Données :**

Les trois fichiers de comptage :

  - **total RNAseq :** *sus_scrofa_linear_counts.tsv*
  - **mRNAseq :** *raffine.gnexpr.readcount.matrix.tsv*
  - **circRNA :** *sus_scrofa_bsj_gene_counts.tsv* => utilisé simplement pour identifier les gènes produisant des circRNAs

**3 échantillons :**

  - *ssc_testis_1* (FR22JFZ201100831) : animal 31
  - *ssc_testis_2* (FR62ND6201203305) : animal 05
  - *ssc_testis_3* (FR79JCW201106154) : animal 54

**Consignes :**

  - extraire les comptages pour les échantillons mentionnés 
  - normaliser les données (paragraphe normalisation : http://genoweb.toulouse.inra.fr/~faraut/circRNAFrontiers)          
  - faire un plot et une régression linéaire pour chaque animal (nuage de points) : comparer l'expression estimée avec du totalRNAseq à celle estimée avec du mRNAseq
  => Réaliser une régression linéaire entre ces estimations (1 par animal)
  - faire apparaître dans le plot les gènes qui produisent des ARNcirc (fichier *sus_scrofa_bsj_gene_counts.tsv*) 
  - montrer que la relation linéaire n'est pas la même pour ces derniers 


# Librairies :

```{r}
library(ggplot2)
library(RColorBrewer)
suppressWarnings(suppressMessages(library(plyr)))
suppressWarnings(suppressMessages(library(dplyr)))
suppressWarnings(suppressMessages(library(ggpmisc)))
suppressWarnings(suppressMessages(library(gridExtra)))
library(edgeR)
```


# 1. Lecture des données :

```{r}
df_linear = read.table("sus_scrofa_linear_counts.tsv", sep="\t", header=TRUE)
df_circrna = read.table("sus_scrofa_bsj_gene_counts.tsv", sep="\t", header=TRUE)
df_readcount = read.table("raffine.gnexpr.readcount.matrix.tsv", sep="\t", header=TRUE)
df_readcount = cbind(gene_id = rownames(df_readcount), df_readcount, row.names = NULL)

df_fr_sc = read.table("junc_ratio/circexplorer/sus_scrofa_junc_ratio_counts.tsv", sep="\t", header=TRUE)
df_fr_sc = ddply(df_fr_sc,"gene_id",numcolwise(sum))

commom_gene_ids = intersect(df_linear$gene_id, df_readcount$gene_id)

# Count per million reads : 

df_linear[,-1] = apply(df_linear[,-1] ,2,function(x){(x/sum(x))*1000000})
df_circrna[,-1] = apply(df_circrna[,-1],2,function(x){(x/sum(x))*1000000})
df_readcount[,-1] = apply(df_readcount[,-1],2,function(x){(x/sum(x))*1000000})
```

# 2. Normalisation des données :

```{r}
# Create a function to normalize data:
normalization <- function(dataframe) {
  Counts = round(data.frame(dataframe[,-1], row.names = dataframe[,1]))
  pseudoCounts <- log2(Counts+1)
  dgeCounts <- DGEList(Counts)

  # We normalize the data in order to take into account the different overall number of reads between samples
  dgeCounts <- calcNormFactors(dgeCounts, method="TMM")
  eff.lib.size <- dgeCounts$samples$lib.size*dgeCounts$samples$norm.factors
  normCounts <- sweep(dgeCounts$counts, 2, eff.lib.size, "/")*10^6
  
  # which is equivalent to normCounts <- cpm(dgeFull)
  pseudoNormCounts <- log2(normCounts + 1)
  df_norm = data.frame(pseudoNormCounts)
  df_norm = cbind(gene_id = rownames(df_norm), df_norm, row.names = NULL)
  return(df_norm)
}

# Apply the normalize function on data:
df_linear_norm = normalization(dataframe = df_linear)
df_circrna_norm = normalization(dataframe = df_circrna)
df_readc_norm = normalization(dataframe = df_readcount)
```


# 3. Extraction des échantillons d'intérêts :

```{r}
df_linear_sp = df_linear_norm[ , c("gene_id", "ssc_testis_1", "ssc_testis_2", "ssc_testis_3")]
colnames(df_linear_sp)<- c("gene_id", "ssc_testis_1_linear","ssc_testis_2_linear","ssc_testis_3_linear")

df_circrna_sp = df_circrna_norm[ , c("gene_id", "ssc_testis_1", "ssc_testis_2", "ssc_testis_3")]
colnames(df_circrna_sp)<- c("gene_id", "ssc_testis_1_circ","ssc_testis_2_circ","ssc_testis_3_circ")

df_readcount_sp = df_readc_norm[ , c("gene_id", "FR22JFZ201100831", "FR62ND6201203305", "FR79JCW201106154")]
colnames(df_readcount_sp)<- c("gene_id", "ssc_testis_1","ssc_testis_2","ssc_testis_3")
colnames(df_readcount_sp)<- c("gene_id", "ssc_testis_1_readc","ssc_testis_2_readc","ssc_testis_3_readc")

df_fr_sc_sp = df_fr_sc[ , c("gene_id", "ssc_testis_1", "ssc_testis_2", "ssc_testis_3")]
colnames(df_fr_sc_sp)<- c("gene_id", "ssc_testis_1_fr","ssc_testis_2_fr","ssc_testis_3_fr")

df_tot = purrr::reduce(list(df_linear_sp, df_circrna_sp, df_readcount_sp, df_fr_sc_sp), left_join, by = "gene_id")

# Delete fr values if NA for circ :
df_tot$ssc_testis_1_fr <- ifelse(is.na(df_tot$ssc_testis_1_circ), NA, df_tot$ssc_testis_1_fr)
df_tot$ssc_testis_2_fr <- ifelse(is.na(df_tot$ssc_testis_1_circ), NA, df_tot$ssc_testis_2_fr)
df_tot$ssc_testis_3_fr <- ifelse(is.na(df_tot$ssc_testis_1_circ), NA, df_tot$ssc_testis_3_fr)

# nrow(df_tot)

df_ssc_1 = df_tot[, c("gene_id", "ssc_testis_1_linear", "ssc_testis_1_readc", "ssc_testis_1_circ", "ssc_testis_1_fr")]
colnames(df_ssc_1)<- c("gene_id", "linear","readc", "circ", "fr")
df_ssc_1 = df_ssc_1[ df_ssc_1$gene_id %in% commom_gene_ids, ]

df_ssc_2 = df_tot[, c("gene_id", "ssc_testis_2_linear", "ssc_testis_2_readc", "ssc_testis_2_circ", "ssc_testis_2_fr")]
colnames(df_ssc_2)<- c("gene_id", "linear","readc", "circ", "fr")
df_ssc_2 = df_ssc_2[ df_ssc_2$gene_id %in% commom_gene_ids, ]

df_ssc_3 = df_tot[, c("gene_id", "ssc_testis_3_linear", "ssc_testis_3_readc", "ssc_testis_3_circ", "ssc_testis_3_fr")]
colnames(df_ssc_3)<- c("gene_id", "linear","readc", "circ", "fr")
df_ssc_3 = df_ssc_3[ df_ssc_3$gene_id %in% commom_gene_ids, ]

# Delete read count == 0 (mRNA) :

df_ssc_1 = df_ssc_1[df_ssc_1$readc != 0, ]
df_ssc_2 = df_ssc_2[df_ssc_2$readc != 0, ]
df_ssc_3 = df_ssc_3[df_ssc_3$readc != 0, ]
```


# 4. Régression linéaire (nuage de points) :

```{r, fig.width=15, fig.height=10}

ggp <- function(dataframe, title) {
  
  circpal <- colorRampPalette(c("saddlebrown", "white"))
  #circpal(10)
  #nocircpal <- colorRampPalette(c("", ""))
  
  dataframe$col = as.factor(as.vector(unlist(lapply(dataframe$circ, function(x){if(is.na(x)){"no circ"} else{"circ"}}))))

  ggplot(dataframe, aes(x=as.numeric(linear), y=as.numeric(readc))) +
 
  geom_point(aes(x=as.numeric(linear), y=as.numeric(readc), color = col)) +
  geom_smooth(method=lm, se=FALSE, fullrange=TRUE, aes(x=as.numeric(linear), y=as.numeric(readc)), colour="dodgerblue2") +
  geom_point(data = subset(dataframe, col == 'circ'), aes(x=as.numeric(linear), y=as.numeric(readc), color = col)) +
  geom_smooth(method=lm, se=FALSE, fullrange=TRUE, data = subset(dataframe, col == 'circ'), aes(x=as.numeric(linear), y=as.numeric(readc)), colour="saddlebrown") +
  labs(title=title, x = "TotalRNAseq expression level (read count/gene norm)",
                    y = "mRNAseq expression level (read count/gene norm)") +
  theme(plot.title = element_text(hjust = 0.5), legend.position="bottom") +
  
  scale_color_manual(values=c('saddlebrown', 'dodgerblue2'), 
                     name="Genes",
                     labels=c("Involved in the production of circRNA", 
                     "Not involved in the production of circRNA")) +
  
  stat_fit_glance(data = subset(dataframe, col == 'no circ'), colour="dodgerblue2",
                  method = "lm", label.x = "left", label.y = 1, size=3.5,
                  method.args = list(formula = y ~ x),
                  mapping = aes(label = sprintf('r^2~"="~%.3f', stat(r.squared))), parse = TRUE) +
  stat_fit_glance(data = subset(dataframe, col == 'circ'), colour="saddlebrown",
                  method = "lm", label.x = "left", label.y = "top", size=3.5,
                  method.args = list(formula = y ~ x),
                  mapping = aes(label = sprintf('r^2~"="~%.3f', stat(r.squared))), parse = TRUE)
}

ggp_1 = ggp(dataframe = df_ssc_1, "ssc_testis_1 (31)")
ggp_2 = ggp(dataframe = df_ssc_2, "ssc_testis_2 (05)")
ggp_3 = ggp(dataframe = df_ssc_3, "ssc_testis_3 (54)")

suppressWarnings(suppressMessages(grid.arrange(ggp_1, ggp_2, ggp_3, ncol = 2, nrow = 2)))
```
**Conclusion :**

```{r, fig.width=15, fig.height=10}

ggp <- function(dataframe, title) {
  
  dataframe$col = as.factor(as.vector(unlist(lapply(dataframe$circ, function(x){if(is.na(x)){"no circ"} else{"circ"}}))))
  
  ggplot(dataframe) +
    
    labs(title=title, x = "TotalRNAseq expression level (read count/gene norm)",
                      y = "mRNAseq expression level (read count/gene norm)") +
    theme(plot.title = element_text(hjust = 0.5), legend.position="bottom") +
    
    geom_point(aes(x=linear, y=readc, color = fr)) +
    geom_smooth(method=lm, se=FALSE, fullrange=TRUE, aes(x=linear, y=readc), colour="grey50") +
    
    geom_point(data = subset(dataframe, is.na(fr) == F), aes(x=linear, y=readc, color = fr)) +
    geom_smooth(method=lm, se=FALSE, fullrange=TRUE, data = subset(dataframe, col == 'circ'), aes(x=linear, y=readc), colour="saddlebrown") +
  
    scale_color_manual(values=c('grey50', 'dodgerblue2'), name="Genes",
                       labels=c("Involved in the production of circRNA", "Not involved in the production of circRNA")) +
    
    scale_color_gradient(low="saddlebrown", high="red") 
}

ggp_1 = ggp(dataframe = df_ssc_1, "ssc_testis_1 (31)")
ggp_2 = ggp(dataframe = df_ssc_2, "ssc_testis_2 (05)")
ggp_3 = ggp(dataframe = df_ssc_3, "ssc_testis_3 (54)")

suppressWarnings(suppressMessages(grid.arrange(ggp_1, ggp_2, ggp_3, ncol = 2, nrow = 2)))
```


```{r, fig.width=15, fig.height=10}

# ggp <- function(dataframe, title) {
#   
#   circpal <- colorRampPalette(c("saddlebrown", "white"))
#   #circpal(10)
#   #nocircpal <- colorRampPalette(c("", ""))
#   
#   dataframe$col = as.factor(as.vector(unlist(lapply(dataframe$circ, function(x){if(is.na(x)){"no circ"} else{"circ"}}))))
# 
#   ggplot(dataframe, aes(x=as.numeric(linear), y=as.numeric(readc))) +
#  
#   geom_point(aes(x=as.numeric(linear), y=as.numeric(readc), color = col)) +
#   geom_smooth(method=lm, se=FALSE, fullrange=TRUE, aes(x=as.numeric(linear), y=as.numeric(readc)), colour="dodgerblue2") +
#   geom_point(data = subset(dataframe, col == 'circ'), aes(x=as.numeric(linear), y=as.numeric(readc), color = col)) +
#   geom_smooth(method=lm, se=FALSE, fullrange=TRUE, data = subset(dataframe, col == 'circ'), aes(x=as.numeric(linear), y=as.numeric(readc)), colour="saddlebrown") +
#   labs(title=title, x = "TotalRNAseq expression level (read count/gene norm)",
#                     y = "mRNAseq expression level (read count/gene norm)") +
#   theme(plot.title = element_text(hjust = 0.5), legend.position="bottom") +
#   
#   scale_color_manual(values=c('saddlebrown', 'dodgerblue2'), 
#                      name="Genes",
#                      labels=c("Involved in the production of circRNA", 
#                      "Not involved in the production of circRNA")) +
#   
#   stat_fit_glance(data = subset(dataframe, col == 'no circ'), colour="dodgerblue2",
#                   method = "lm", label.x = "left", label.y = 1, size=3.5,
#                   method.args = list(formula = y ~ x),
#                   mapping = aes(label = sprintf('r^2~"="~%.3f', stat(r.squared))), parse = TRUE) +
#   stat_fit_glance(data = subset(dataframe, col == 'circ'), colour="saddlebrown",
#                   method = "lm", label.x = "left", label.y = "top", size=3.5,
#                   method.args = list(formula = y ~ x),
#                   mapping = aes(label = sprintf('r^2~"="~%.3f', stat(r.squared))), parse = TRUE)
# }
# 
# ggp_1 = ggp(dataframe = df_ssc_1, "ssc_testis_1 (31)")
# ggp_2 = ggp(dataframe = df_ssc_2, "ssc_testis_2 (05)")
# ggp_3 = ggp(dataframe = df_ssc_3, "ssc_testis_3 (54)")
# 
# suppressWarnings(suppressMessages(grid.arrange(ggp_1, ggp_2, ggp_3, ncol = 2, nrow = 2)))
```






